### PR TITLE
resolve: refactor path resolution

### PR DESCRIPTION
**Pull request by jseyfried**
This is a pure refactoring, modulo minor diagnostics improvements.
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc @eddyb @petrochenkov

**Comment by petrochenkov** [comment]
Hey, I have a relatively large rewrite of high level path resolution and error reporting in progress: https://github.com/petrochenkov/rust/tree/altname.
I'll look through this patch and see how they integrate (it may make sense to wait with changes lying above `resolve_path`/`resolve_possibly_assoc_item` in call stack).

**Comment by petrochenkov** [comment]
Reviewed. The changes are mostly orthogonal to my branch, the main contention is about `resolve_path` signature. The merge may be a bit painful, but nothing critically bad.
The patch looks good to me overall.

File: src/librustc/hir/lowering.rs [hunk 1]
```diff
@@ -71,13 +71,13 @@ pub struct LoweringContext<'a> {
 
 pub trait Resolver {
     // Resolve a global hir path generated by the lowerer when expanding `for`, `if let`, etc.
-    fn resolve_generated_global_path(&mut self, path: &hir::Path, is_value: bool) -> Def;
+    fn resolve_hir_path(&mut self, path: &hir::Path, is_value: bool) -> PathResolution;
```
Comment (eddyb) [inline src/librustc/hir/lowering.rs]: FWIW, changes in this module conflict with #37676, so you can either revert them to merge earlier, or wait for that PR (which would be merged already if it weren't for windows).
Also keep in mind that `PathResolution` doesn't really matter after HIR lowering in #37676, but YMMV.

**Comment by eddyb** [comment]
@nrc See https://github.com/rust-lang/rust/pull/38014#discussion_r89680484. @bors r-

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -191,10 +189,6 @@ fn resolve_struct_error<'b, 'a: 'b, 'c>(resolver: &'b Resolver<'a>,
                                         span: syntax_pos::Span,
                                         resolution_error: ResolutionError<'c>)
                                         -> DiagnosticBuilder<'a> {
-    if !resolver.emit_errors {
-        return resolver.session.diagnostic().struct_dummy();
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: There *should* be a way to do speculative resolution without reporting errors.
It's used at least once now in fresh binding disambiguation and I'm using it in my error reporting branch as well.

Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: @petrochenkov You can still do speculative resolution by passing `record_used: None` to `resolve_path`, `resolve_ident_in_lexical_scope`, or `resolve_name_in_module`.

**Comment by jseyfried** [comment]
Rebased and added a new commit -- r? @eddyb for new commit.

File: src/librustc/hir/lowering.rs [hunk 3]
```diff
@@ -353,9 +350,6 @@ impl<'a> LoweringContext<'a> {
             // `<I as Iterator>::Item::default`.
             let ty = self.ty(p.span, hir::TyPath(hir::QPath::Resolved(qself, path)));
 
-            // Associate that innermost path type with the base Def.
-            self.resolver.record_resolution(ty.id, resolution.base_def);
-
             ty
```
Comment (eddyb) [inline src/librustc/hir/lowering.rs]: You can avoid the `ty` variable.

**Review by eddyb — APPROVED** [review]
r=me with nit fixed

**Pull request merged**